### PR TITLE
fix: Add test for nushell integration and fix some bugs

### PR DESCRIPTION
**Pull request by fwfurtado**
# Summary

Fix bug of using `first`  when hasn't a value for an expression
Fix support for `ASDF_DIR` env var

Add tests for nushell integration

Closes: #1382

### CODE REVIEW

**Comment by Stratus3D** [comment]
Thanks for the PR! It looks like tests are failing on MacOS 10.15. Can you take a look?

Also, is rust required for nushell? Is there a way to install it without rust? I would like to keep the number of CI dependencies to a minimum.

**Comment by fwfurtado** [comment]
> Thanks for the PR! It looks like tests are failing on MacOS 10.15. Can you take a look?
> 
> Also, is rust required for nushell? Is there a way to install it without rust? I would like to keep the number of CI dependencies to a minimum.

No problem! I've fixed the failing test on MacOS 10.15 a while ago.

Unfortunately the ubuntu hasn't a package of nushell, and according installation guide we need to download the nushell via **cargo**

**Comment by fwfurtado** [comment]
> Thanks for the PR! It looks like tests are failing on MacOS 10.15. Can you take a look?
> 
> Also, is rust required for nushell? Is there a way to install it without rust? I would like to keep the number of CI dependencies to a minimum.

No problem! I've fixed the failing test on MacOS 10.15 a while ago.

Unfortunately the ubuntu hasn't a package of nushell, and according installation guide we need to download the nushell via **cargo**

File: test/asdf_nu.bats [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    let-env ASDF_NU_DIR = '$PWD'
+
+    source asdf.nu 
+
+    echo \$env.ASDF_DIR
+  ")
+
+  [ "$?" -eq 0 ]
+  output=$(echo "$result" | grep "asdf")
+  [ "$output" != "" ]
```
Comment (Stratus3D) [inline test/asdf_nu.bats:27]: Would it be possible to assert that the path is correct here?

File: test/asdf_nu.bats [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    let-env PATH = ( '$(cleaned_path)' | split row ':' )
+    let-env ASDF_NU_DIR = '$PWD'
+
+    source asdf.nu 
+
+    echo \$env.PATH
+ ")
+  [ "$?" -eq 0 ]
+  output=$(echo "$result" | grep "asdf")
+  [ "$output" != "" ]
```
Comment (Stratus3D) [inline test/asdf_nu.bats:43]: Same as above.

**Comment by Stratus3D** [comment]
I'm looking at the runtimes of the tests on Ubuntu I'm seeing them increase from around 2 minutes on latest master to about 8 minutes on this PR branch. I'm not sure I want to increase build times that much just so we can have tests for nushell. Perhaps we could just limit them to running on MacOS where nushell install is fast?

File: asdf.nu [hunk 3]
```diff
@@ -1,26 +1,18 @@
-let possible_dirs = [ 
-                        ($env | get --ignore-errors ASDF_DIR | default '' | first) , 
-                        '/opt/asdf-vm/asdf.nu',
```
[the remark this answers is not in the document]
Comment (fwfurtado) [reply asdf.nu:3]: This is not needed anymore once this was an attempt to predict the installation path. However when I realized how fish and elvish do this thing I decided to change this to use the `ASDF_DIR` env var instead

**Comment by fwfurtado** [comment]
> I'm looking at the runtimes of the tests on Ubuntu I'm seeing them increase from around 2 minutes on latest master to about 8 minutes on this PR branch. I'm not sure I want to increase build times that much just so we can have tests for nushell. Perhaps we could just limit them to running on MacOS where nushell install is fast?

@Stratus3D I'd change the ubuntu tests to use a binary version of nushell. 
Can you check if this build/test time is acceptable or if we need to disable nushell's tests for Ubuntu?

Although, I've needed to skip tests in Ubuntu 18.04 becase this don't support minimal  version of GCLib required for nushell latest.

File: test/asdf_nu.bats [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+setup() {
+  cd $(dirname "$BATS_TEST_DIRNAME")
+}
+
+cleaned_path() {
+  echo $PATH | tr ':' '\n' | grep -v "asdf" | tr '\n' ':'
+}
+
+@test "exports ASDF_DIR" {
+  if [ "${SKIP_NUSHELL_TESTS}" = 'YES' ]; then
```
Comment (Stratus3D) [inline test/asdf_nu.bats:14]: Is it not possible to skip the entire file with a single check at the top?

e.g. something like this:

```
if [ "${SKIP_NUSHELL_TESTS}" = 'YES' ]; then
  echo "skipping nushell tests"
  exit 0
fi
```

File: .github/workflows/tests.yml [hunk 5]
```diff
@@ -56,3 +67,4 @@ jobs:
         run: bats test
         env:
           GITHUB_API_TOKEN: ${{ github.token }}
+          SKIP_NUSHELL_TESTS:  "${{ matrix.os == 'ubuntu-18.04' && 'YES' || 'NO' }}"
```
Comment (Stratus3D) [inline .github/workflows/tests.yml:70]: The `|| 'NO'` is probably unnecessary since you explicitly check for `YES` in the test code.

**Comment by Stratus3D** [comment]
Left two more comments, otherwise looks great!

**Comment by Stratus3D** [comment]
I guess my comment about exiting early from the nushell tests was wrong. Now it's saying the tests have failed when they are skipped. I did find this on SO - https://stackoverflow.com/questions/49991275/is-there-a-way-to-bail-from-a-bats-test

File: test/asdf_nu.bats (hunk 4 above)
Comment (fwfurtado) [reply test/asdf_nu.bats:14]: According to [FAQ](https://bats-core.readthedocs.io/en/stable/faq.html#how-can-i-exclude-all-tests-of-a-file-from-a-test-run) of bats it is not possible.

When I tried this the test was broken.

**Comment by fwfurtado** [comment]
> I guess my comment about exiting early from the nushell tests was wrong. Now it's saying the tests have failed when they are skipped. I did find this on SO - https://stackoverflow.com/questions/49991275/is-there-a-way-to-bail-from-a-bats-test

I moved the skip statement to `setup` function

**Comment by Stratus3D** [comment]
Thanks for addressing all the comments @fwfurtado ! 🚀

**Review by Stratus3D — APPROVED** [review]

**Pull request merged**